server/core_storage.go

```diff
@@ -647,24 +646,27 @@ func storagePrepBatch(batch *pgx.Batch, authoritativeWrite bool, op *StorageOpWr
 		// allows us to fetch row state after update even if update itself fails WHERE
 		// condition.
 		// That is returned values are final state of the row regardless of UPDATE success
+		// Order by is necessary as UNION ALL may not guarantee ordering of results, but it is required
+		// to ensure that if update happens values from said update are returned.
+		// At most 2 rows are ever returned so ordering operation should be negligible.
 		query = `
 		WITH upd AS (
 			UPDATE storage SET value = $4, version = $5, read = $6, write = $7, update_time = now()
 			WHERE collection = $1 AND key = $2 AND user_id = $3 AND version = $8
 		` + writeCheck + `
-			AND NOT (storage.version = $5 AND storage.read = $6 AND storage.write = $7) -- micro optimization: don't update row unnecessary
 			RETURNING read, write, version, create_time, update_time
 		)
-		(SELECT read, write, version, create_time, update_time from upd)
+		(SELECT read, write, version, create_time, update_time, true AS update FROM upd)
 		UNION ALL
-		(SELECT read, write, version, create_time, update_time FROM storage WHERE collection = $1 and key = $2 and user_id = $3)
+		(SELECT read, write, version, create_time, update_time, false AS update FROM storage WHERE collection = $1 and key = $2 and user_id = $3)
+		ORDER BY update DESC
```

Comment:
This variant of the query doesn't invoke unnecessary read of the same key and still produces correct result even if `UNION ALL` brances happen to be executed out of order:


```
WITH upd AS (
			UPDATE storage SET value = $4, version = $5, read = $6, write = $7, update_time = now()
			WHERE collection = $1 AND key = $2 AND user_id = $3 AND version = $8
		` + writeCheck + `
			RETURNING read, write, version, create_time, update_time
		)
(SELECT read, write, version, create_time, update_time, true AS update FROM upd WHERE EXISTS (SELECT FROM upd))
UNION ALL
(SELECT read, write, version, create_time, update_time, false AS update FROM storage WHERE collection = $1 and key = $2 and user_id = $3 AND NOT EXIST (SELECT FROM upd)
LIMIT 1`
```

It works by making `UNION ALL` branches mutually exclusive with `EXISTS`/`NOT EXISTS` check on `upd` CTE. Data-modifying CTEs are documented to be executed only once:

> Data-modifying statements in WITH are executed exactly once, and always to completion, independently of whether the primary query reads all (or indeed any) of their output. 

So referencing CTE multiple times wont create multiple attempts to modify data and therefore cheap and safe.